src/basicfuns.jl

```diff
@@ -311,10 +311,20 @@ $(SIGNATURES)
 Return `log(exp(x) + exp(y))`, avoiding intermediate overflow/undeflow, and handling
 non-finite values.
 """
-function logaddexp(x::Real, y::Real)
-    # ensure Δ = 0 if x = y = ± Inf
-    Δ = x == y ? zero(x - y) : abs(x - y)
-    max(x, y) + log1pexp(-Δ)
+logaddexp(x::Real, y::Real) = _logaddexp(map(float, promote(x, y))...)
+
+function _logaddexp(x::T, y::T) where T <: Real
+    if (isnan(x) || isnan(y)) return T(NaN) end
+
+    if x < y
+        diff = x - y
+        maxval = y
+    else
+        diff = y - x
+        maxval = x
+    end
+
+    isnan(diff) ? maxval : maxval + log1pexp(diff)
```

Comment:
I'm not completely happy about the additional internal method, I don't think it should be needed here since we don't dispatch on it or reuse it in other function calls. Maybe this could be simplified to
```suggestion
function logaddexp(x::Real, y::Real)
    # Compute max = Base.max(x, y) and diff = x == y ? zero(x - y) : abs(x - y)
    # in a faster type-stable way
    a, b = promote(x, y)
    if a < b
        diff = b - a
        max = b
    else
        # ensure diff = 0 if a = b = ± Inf
        diff = a == b ? zero(a - b) : a - b
        max = !isnan(b) ? a : b
    end

    return max + log1pexp(diff)
```
?

In my benchmarks this leads to a similar improvement (BTW it's not necessary to use broadcasting in benchmarking and I also wonder if your benchmarks are off since you did not interpolate `x` and `y` but used global variables):
```julia
julia> @btime logaddexp(x, y) setup=(x = 1000 * randn(); y = 1000 * randn());
  3.686 ns (0 allocations: 0 bytes)

julia> @btime logaddexp_new(x, y) setup=(x = 1000 * randn(); y = 1000 * randn());
  2.612 ns (0 allocations: 0 bytes)

julia> @btime logaddexp.(x, y) setup=(x = 1000 * randn(10_000); y = 1000 * randn(10_000));
  125.475 μs (2 allocations: 78.17 KiB)

julia> @btime logaddexp_new.(x, y) setup=(x = 1000 * randn(10_000); y = 1000 * randn(10_000));
  82.540 μs (2 allocations: 78.17 KiB)
```

Reply:
Yes, actually we can handle the type without resorting to yet another function. Some tests failed though with your proposed version so I kept the current "branching version" .

Reply:
> Some tests failed though with your proposed version so I kept the current "branching version" .

Seems I mistyped `diff` instead of `-diff` - then all tests pass :smile: